v1.2.0 - findAll, ImportPath class and use of ascii_art_tree.

gmpassos opened the pull request:
- Moved code to class `ImportPath`:
  - Allows integration with other packages.
  - Facilitates tests.
- Using `ascii_art_tree` to show the output tree, with styles `dots` (original) and `elegant`.
- CLI:
  - Added options:
    - `--regexp`: to use `RegExp` to match the target import.
    - `--all`: to find all the import paths. 
    - `-q`: for a quiet output (only displays found paths).
    - `-s`: strips the search root directory from displayed import paths.
    - `--elegant`: use `elegant` style for the output tree.
    - `--dots`: use `dots` style for the output tree.
  - Improved help with examples.
- Updated `README.md` to show CLI and Library usage.
- Added tests and coverage (80%).
- Added GitHub Dart CI.

- Updated dependencies compatible with Dart `2.14.0` (was already dependent to SDK `2.14.0`): 
  - sdk: '>=2.14.0 <3.0.0'
  - package_config: ^2.1.0
  - path: ^1.8.3
  - ascii_art_tree: ^1.0.2
  - lints: ^1.0.1
  - dependency_validator: ^3.2.2
  - test: ^1.21.4
  - coverage: ^1.2.0

jakemac53 commented on lib/src/import_path_base.dart:
```diff
@@ -0,0 +1,283 @@
+  }
+
+  List<List<String>> _searchImportPaths(
+    Uri node, {
```
Can we give this a better name, maybe `current`?

gmpassos replied:
You mean for the `node` parameter? Maybe `currentNode`

jakemac53 replied:
Yeah that sounds fine to me

jakemac53 commented on lib/src/import_path_base.dart:
```diff
@@ -0,0 +1,283 @@
+  final Uri from;
+
+  /// The import to find. Can be an [Uri] or a [RegExp].
+  dynamic importToFind;
```
Actually I think it would be worth creating an abstraction here, something like an `ImportFinder` class (or function type). Then you can have Uri and RegExp implementations of that, and we can have a proper type for this field both here and in the actual class, and avoid the repeated `is` checks.

jakemac53 commented on lib/src/import_path_base.dart:
```diff
@@ -0,0 +1,283 @@
+
+    if (filePath == null) {
+      if (!quiet) {
+        printMessage('» [WARNING] Unable to resolve Uri $uri, skipping it');
```
We probably always want to print this warning as it indicates an error.

gmpassos replied:
This only happens if the current working directory is not the dart project directory. Try to scan package `foo` from the `import_path` directory:

```shell
cd workspace/import_path

dart run bin/import_path.dart ../foo/web/main.dart dart:html
```

jakemac53 replied:
It could happen if you haven't ran `pub get` since you added some package to  your pubspec, but also since you described the use case for quiet mode above I think we do want to avoid logging it in this mode.

jakemac53 commented on lib/src/import_path_base.dart:
```diff
@@ -0,0 +1,283 @@
+    for (var import in newImports) {
+      if (_matchesImport(importToFind, import)) {
+        var foundPath = [...parents, import.toString()];
+        found.add(foundPath);
```
Consider creating a ChildNode type, which has a pointer to its parent node (but not the other way).  This should avoid the need for lists entirely for the paths, which will avoid extra copying around like this. You could instead return iterables, since it's easy to crawl from a child up to all its parents.

We could actually represent a more complex graph that way too, if we allowed multiple parents. And possibly have a mode where we would serve a visual graph or output a graph in dot format? (not for now, but could be a nice follow up).

A representation like that would also resolve the issue I described above, finding an already "walked" node would actually mean just adding the current node to its parents.

gmpassos replied:
Fixed with the use of package graph_explorer

jakemac53 replied:
Fwiw, I tested this using my fake package generator tool, which basically just generates packages with really complex import graphs for worst case scenarios. It seems to perform just fine in the shortest path mode 👍.

When there are many paths the `--all` mode definitely struggles a lot (took several minutes in my case, I gave up after 5). I think there could definitely be a lot of optimization there - ultimately the paths are still being represented with separate lists from what I could tell, which involves a huge amount of duplicate work. They should be represented as Iterables instead, which can be lazily printed and constructed.

I think the model for output in that mode should also really be a graph format (something like the DOT language for graphviz), and we could even then do something similar to the [pubviz](https://pub.dev/packages/pubviz) package does, creating/serving an HTML page for viewing in a browser. The console is just really not a good mechanism for efficiently visualizing large graphs 🤣. In general this format would also be more useful for tools than JSON (for the same reasons, orders of magnitude smaller representation).

I could also envision possibly some sort of interactive console mode, where you can say "show me the next path". If we can lazily materialize the paths, then we can always only show the first X (maybe 10?) and then have a sort of paginated setup where you can ask for more. That way it won't just hang forever, or spit out an infinite stream of crud to the screen :).

This should all be a follow up if anything, for many cases the existing `--all` implementation will probably work well enough. But I do think it should possibly print a warning that on large programs it might not complete in reasonable time.

gmpassos replied:
Can I test your generator? So I can try to find the bottleneck in the code, beside that parents list?

jakemac53 replied:
https://gist.github.com/jakemac53/890f7dac88419f176e964458f9a12f7e

But, this is a fundamental flaw that you won't be able to optimize yourself out of I am quite confident :).

gmpassos commented on .github/workflows/dart.yml:
```diff
@@ -0,0 +1,95 @@
+      - name: Install dependencies
+        run: dart pub get
+      - name: Upgrade dependencies
+        run: dart pub upgrade
```
In theory I agree with you, until I had some real world issues.

I only add the upgrade command because in some cases the GitHub workflow has come cached packages and is not getting the last version for some pre-installed packages.

Dart pub get only gets the last version if it's downloading the package for the first time.

gmpassos replied:
`dart pub get`

`If the system cache doesn’t already contain the dependencies, dart pub get updates the cache, downloading dependencies if necessary.`

From:
https://dart.dev/tools/pub/cmd/pub-get

When running `dart pub get`, if a dependency is already in the cache and matches the constraint, it won't download a newer version.

jakemac53 replied:
That isn't my experience with how it works, but I am fine with using `dart pub upgrade`. We definitely don't need both though (just remove the `get` then).

gmpassos commented on .github/workflows/dart.yml:
```diff
@@ -0,0 +1,95 @@
+        run: dart analyze --fatal-infos --fatal-warnings .
+      - name: dependency_validator
+        run: dart run dependency_validator
+      - name: dart pub publish --dry-run
```
This only checks the basic, to avoid any issue in `pubspec` for a package.

gmpassos replied:
Can we keep it? I prefer to maintain this basic check, so any from also checks the pubspec.

gmpassos commented on lib/src/import_path_base.dart:
```diff
@@ -0,0 +1,283 @@
+
+  /// The function to print messages/text. Default: [print].
+  /// Called by [printMessage].
+  void Function(Object? m) messagePrinter;
```
This is to allow a custom `messagePrinter`. It's used in tests and I need it to integrate with our package.

gmpassos commented on lib/src/import_path_base.dart:
```diff
@@ -0,0 +1,283 @@
+
+  /// Resolves the [searchRoot] using [from] parent.
+  /// See [commonRootDirectories].
+  String resolveSearchRoot() {
```
If public it can be customized by an implementation that extends `ImportPath` (we are integrating `ImportPath` in our tool)

jakemac53 replied:
Can you explain more what the point of this actually is? In particular with regards to `commonRootDirectories`?

This seems like new logic but I am having a very hard time following the intentions of the code.

If the goal is to find the root of the package that `from` lives in, we could use the package config to better do that and show paths relative to that package root without the assumptions we have here?

gmpassos commented on lib/src/import_path_base.dart:
```diff
@@ -0,0 +1,283 @@
+  /// Return the search root to [strip] from the displayed import paths.
+  /// If [strip] is `false` returns `null`.
+  /// See [searchRoot].
+  String? get stripSearchRoot => strip ? searchRoot : null;
```
Same, the implementation can be extended. Let's say that a complex `stripSearchRoot` is needed by some tool.

gmpassos commented on bin/import_path.dart:
```diff
@@ -2,105 +2,69 @@
+  --regexp   # Parses `%targetImport` as a `RegExp`.
+  --all      # Searches for all the import paths.
+  -s         # Strips the search root directory from displayed import paths.
+  -q         # Quiet output (only displays found paths).
```
The default is to not show any warning?

The actual levels are:

- quiet
- standard
- verbose

There's no verbose mode. The quiet is to ensure that some program that is reading the `import_path` output is only reading the path tree.

gmpassos commented on lib/src/import_path_base.dart:
```diff
@@ -0,0 +1,283 @@
+  /// Performs the imports search and returns the tree.
+  /// - If [dots] is `true` uses the `dots` style for the tree.
+  /// - See [ASCIIArtTree].
+  Future<ASCIIArtTree?> search({bool dots = false}) async {
```
If we want to integrate the class `ImportPath` in out tool, we don't want the output text, we want the result of the `search` method

gmpassos commented on lib/src/import_path_base.dart:
```diff
@@ -0,0 +1,283 @@
+    if (!findAll) {
+      foundPaths.sort((a, b) => a.length.compareTo(b.length));
+      var shortest = foundPaths.first;
+      return [shortest];
```
The idea was to use the same code for shortest path and all paths. But the optimized code is small, and we can also have the optimized version too.

jakemac53 replied:
We could have shared code still (but this algorithm wouldn't work). In particular an iterative algorithm makes it much easier to bail out early (and we would need to do a BFS instead of a DFS so we know we hit the shortest path first and then its just a flag to immediately bail on the first solution found).

gmpassos replied:
I will reimplement the search algorithm to be BFS and not recursive, producing all the matching paths of the import graph...

gmpassos commented on CHANGELOG.md:
```diff
@@ -1,3 +1,32 @@
+    - `--all`: to find all the import paths. 
+    - `-q`: for a quiet output (only displays found paths).
+    - `-s`: strips the search root directory from displayed import paths.
+    - `--elegant`: use `elegant` style for the output tree.
```
Done:

- Added `--format` option.
- Implemented `ImportPathStyle` (dots, elegant, json).
- Added the JSON output.

gmpassos commented:
I have made the following changes:

- Moved code to class `ImportPath`, `ImportPathScanner` and `ImportParser`:
- `ImportParser`:
  - Added faster parser.
  - Added support to conditional imports.
- `ImportPathScanner`:
  - Added support to final all the target imports.
  - Optimized resolution of paths graph.
- CLI:
  - `--format`: Defines the style for the output tree (elegant, dots, json).
  - `--fast`: to enable a fast import parser.
- ascii_art_tree: ^1.0.5
- graph_explorer: ^1.0.0
  - Using this package to perform graph operations and path scan.

gmpassos commented on lib/src/import_path_base.dart:
```diff
@@ -0,0 +1,283 @@
+    if (walked == null) {
+      walked = {node};
+    } else if (!walked.add(node)) {
+      return found;
```
Fixed with the use of package `graph_explorer `

gmpassos commented on lib/src/import_path_base.dart:
```diff
@@ -0,0 +1,283 @@
+    }
+  }
+
+  List<List<String>> _searchImportPaths(
```
Done: Fixed with the use of package `graph_explorer`

jakemac53 commented on lib/src/import_path_base.dart:
```diff
@@ -0,0 +1,276 @@
+    var rootDirName = p.split(rootPath).last;
+
+    if (commonRootDirectories.contains(rootDirName)) {
+      var rootPath2 = p.dirname(rootPath);
```
Lets give this a better name, `rootPathParent`?

jakemac53 commented on lib/src/import_path_base.dart:
```diff
@@ -0,0 +1,276 @@
+  /// Performs the imports search and returns the tree.
+  /// - [style] defines [ASCIIArtTree] style. Default: [ASCIIArtTreeStyle.elegant]
+  /// - See [searchPaths] and [ASCIIArtTree].
+  Future<ASCIIArtTree?> search(
```
I don't think this method should be on this class - at most it could be an extension method? I don't think its something that should really be overridden etc, and it's tightly coupling this class to the ASCIIArtTree itself which isn't a good separation of concerns. 

I would probably rename `searchPaths` to just `search`, and drop this (its only used in one place and trivial to inline there).

jakemac53 commented on lib/src/import_path_base.dart:
```diff
@@ -0,0 +1,276 @@
+  /// Executes the import search and prints the results.
+  /// - [style] defines the output format. Default: [ImportPathStyle.elegant]
+  /// - See [search], [printMessage] and [ASCIIArtTree].
+  Future<ASCIIArtTree?> execute(
```
I also feel like this could be an extension method, or just inlined into the `bin/` script.

It is also weird that it both returns the `ASCIIArtTree` and prints out the paths, but not necessarily in ascii format.

jakemac53 commented on lib/src/import_path_base.dart:
```diff
@@ -0,0 +1,276 @@
+  /// Performs the imports search and returns the found import paths.
+  /// - [packageDirectory] is used to resolve . Default: [Directory.current].
+  /// - Uses [ImportPathScanner].
+  Future<List<List<String>>> searchPaths({Directory? packageDirectory}) async {
```
Mentioned above but I do really think that List<List<String>> is the wrong type here. Given this is a public API I am a bit hesitant to expose that.

You could type it as `Future<Iterable<Iterable<String>>>` which would give more flexibility to change it to lazy iterables later on?

gmpassos replied:
I vote for `Future<Iterable<List<String>>>`, since `List<String>` is a path, won't be possible to be lazy.

jakemac53 replied:
The path can be lazy though - that is what I am trying to say :). But it cannot be if it is represented by a List.

It would certainly take some work to get there, but fundamentally each path can be created just by doing a traversal of nodes. I am not sure I can fully explain what I have in mind over github comments though lol 🤣 .

We could also just consider actually exposing the graph representation more directly? If you were given the root of the tree, with only the edges that actually lead to the URI that you are searching for present, that would be a compact representation that you could pretty easily convert into anything else.

jakemac53 commented on lib/src/import_path_parser.dart:
```diff
@@ -0,0 +1,190 @@
+
+/// Dart import parser.
+/// Parses for import directives in Dart files.
+class ImportParser extends ImportWidget {
```
I can't see any difference using the faster parser - and it's a lot of extra stuff to maintain. Ultimately, its only going to be a constant performance factor improvement and we are dealing with something exponential in its complexity, so a constant factor doesn't help. This implementation still looks at the whole file anyways though.

So I think we should drop the fast parser mode (or if you want, you could make it injectable or something and have your own implementation?). I don't want to support this parser ultimately, unless it made a really large difference in performance.

gmpassos replied:
Testing with 2 real projects, the difference with the faster parser is significant. It's 3x faster.

jakemac53 replied:
That makes sense that its more noticeable in a real project, I realized that my pet project libraries only have directives 🤣 .

But, while it might be faster at parsing, that isn't overall the slow thing here, and I really don't want to own/maintain a regex based parser.

If we just allow this class to accept an ImportParser (which should really only do String -> dependencies conversion), then you could own this regex based one and use it in your own scenarios?

gmpassos replied:
Well, it's a very simple `RegExp`. The code finds the last import, then there's a fallback to the full parser in case of error. The idea is not to give a perfect fast parser, just an alternative.

We are using only the fast parser here for out projects. (already using the tool here).

gmpassos replied:
FYI:

https://github.com/dart-lang/sdk/issues/53498

jakemac53 replied:
> Well, it's a very simple `RegExp`. The code finds the last import, then there's a fallback to the full parser in case of error. The idea is not to give a perfect fast parser, just an alternative.
> 
> We are using only the fast parser here for out projects. (already using the tool here).

But it will still have to be updated as the language evolves (we have multiple proposals out currently that could affect this, they may not land, but I don't have a lot of time to spend maintaining this).

> FYI:
> 
> [dart-lang/sdk#53498](https://github.com/dart-lang/sdk/issues/53498)

I definitely think it would be great if we could get some better support from the SDK here, we do this in a lot of places.

gmpassos commented on lib/src/import_path_base.dart:
```diff
@@ -0,0 +1,276 @@
+/// Base class [ImportPath], [ImportPathScanner] and [ImportParser].
+abstract class ImportWidget {
+  /// If `true`, we won't call [printMessage].
+  final bool quiet;
```
`messagePrinter` is not a logging mechanism, it's an output mechanism.
